agent_build_refactored/docker_image_builders.py

```diff
@@ -0,0 +1,771 @@
+                "agent_build/requirement-files/docker-image-requirements.txt",
+                "agent_build/requirement-files/compression-requirements.txt",
+                "agent_build/requirement-files/main-requirements.txt",
+            ],
+            environment_variables={
+                "PYTHON_BASE_IMAGE": base_distro.python_base_image,
+                "DISTRO_NAME": base_distro.name,
+                "RESULT_IMAGE_TARBALL_NAME": f"{base_image_name_prefix}:{image_platform.to_dashed_str}.tar",
+                "PLATFORM": str(image_platform),
+                "COVERAGE_VERSION": TEST_IMAGE_COVERAGE_VERSION,
+            },
+            github_actions_settings=GitHubActionsSettings(
+                cacheable=True, pre_build_in_separate_job=True
+            ),
+        )
+
+
+class BaseImageBuilderStep(ArtifactRunnerStep):
+    """
+    This step combines all platform-specific images  from its required steps into one multi-arch base image
+    """
+
+    def __init__(
+        self,
+        base_distro: BaseDistroSpec,
+        supported_platforms: List[DockerPlatformInfo],
+    ):
+        """
+        :param base_distro: Spec of the base distro with all its needed information.
+        :param supported_platforms: List of platforms that has to be supported by the final image.
+            For each platform, this step creates separate platform-specific "child" step.
+        """
+        self.base_image_name_prefix = f"agent-base-image-{base_distro.name}"
+
+        required_steps = {}
+        for plat in supported_platforms:
+            platform_step = BaseImagePlatformBuilderStep(
+                base_distro=base_distro,
+                image_platform=plat,
+                base_image_name_prefix=self.base_image_name_prefix,
+            )
+
+            required_steps[f"_BASE_PLATFORM_STEP_{platform_step.name}"] = platform_step
+
+        super(BaseImageBuilderStep, self).__init__(
+            name=self.base_image_name_prefix,
+            script_path="agent_build_refactored/docker/steps/build_base_image.py",
+            tracked_files_globs=[
+                # Also add module files which are used by script.
+                "agent_build_refactored/tools/steps_libs/container.py",
+                "agent_build_refactored/tools/steps_libs/subprocess_with_log.py",
+                "agent_build_refactored/tools/steps_libs/build_logging.py",
+                "agent_build_refactored/tools/steps_libs/constants.py",
+            ],
+            required_steps=required_steps,
+            environment_variables={
+                "BASE_IMAGE_NAME_PREFIX": self.base_image_name_prefix,
+            },
+            github_actions_settings=GitHubActionsSettings(
+                cacheable=True,
+            ),
+        )
+        self.base_distro = base_distro
+        self.supported_platforms = supported_platforms
+
+    def __post_init__(self):
```

Comment:
Do we need this method for anything? I don't see it being overridden anywhere, but maybe I missed it.

Reply:
woops, sorry. It was forgotten when I tried to make ``RunnerStep`` as a ``dataclass``. Didn't worked well since this class is to big, and "dataclass-style" declaration of it is less readable

Reply:
PS: The main idea was to make a steps dataclass `frozen` to ensure that checksum calculation for cache  can not be somehow mutated.